### PR TITLE
Initial Implementation

### PR DESCRIPTION
**Pull request by mattruggio**
This PR contains the initial implementation.  It allows for the declaration and creation of:

1. Text files
2. Zip archives

There are two methods for extension:

1. Register your own document types (what and how files get generated)
2. Provide your own Store implementation (where the files ultimately get moved to)

### CODE REVIEW

File: spec/file_helper.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+#
+
+require 'yaml'
+
+def read_yaml_file(*filename)
+  YAML.safe_load(read_binary_file(*filename))
+end
+
+def read_binary_file(*filename)
+  File.open(File.join(*filename), 'rb').read
```
Comment (gerryster) [inline spec/file_helper.rb:17]: Why open the YAML file in binary mode?

**Review by gerryster — CHANGES_REQUESTED** [review]
Most of my comments are fairly minor and several are regarding documentation. One overall thought is the creation of zip files without an enclosing directory. This is a pet peeve of mine. When I extract a zip file which puts its contents in the current directory instead of extracting everything in an enclosing directory, I am irritated about the clutter that was just created. It would be nice if one could specify an enclosing directory for zip files via the blueprint. A similar concern is that even though there is file uniqueness checking at the top level, it would be possible to create a zip file that when extracted, conflicts with other files. For example, modifying your hello_worlds_zip.yaml example:

```
documents:
  - type: text
    filename: hello_world_1.txt
    data: FIRST HELLO WORLD!
  - type: text
    filename: hello_world_2.txt
    data: SECOND HELLO WORLD!
  - type: zip
    filename: nested.zip
    blueprint:
      documents:
        - type: text
          filename: hello_world_1.txt
          data: THIRD HELLO WORLD!
```

Note that I do not get an error about hello_world_1.txt being a duplicate file name. Typically unzip programs warn a user before overwriting a file when extracting. However, it would probably be handy to have some guardrails for Blueprint authors to prevent them from creating scenarios where the end user of the created files has to make an unexpected decision about duplication.

File: lib/file_composer/documents/zip.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      def initialize(filename:, blueprint: {})
+        super(filename: filename)
+
+        @blueprint = Blueprint.make(blueprint)
+      end
+
+      def write!(temp_root = '', store = Stores::Null.new)
+        results               = blueprint.write!(temp_root)
+        total_time_in_seconds = results.sum(&:time_in_seconds)
+        physical_filename     = zip!(temp_root, results)
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/file_composer/documents/zip.rb:27]: I went back and forth on this one.

File: lib/file_composer/stores/local.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        @date = date
+        @root = root.to_s
+
+        freeze
+      end
+
+      def move!(filename)
+        make_final_filename(filename).tap do |final_filename|
+          ensure_directory_exists(final_filename)
+          FileUtils.cp(filename, final_filename)
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/file_composer/stores/local.rb:26]: Good call

File: lib/file_composer/documents/file_result.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+#
+# Copyright (c) 2020-present, Blue Marble Payroll, LLC
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+#
+
+module FileComposer
+  class Documents
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply lib/file_composer/documents/file_result.rb:11]: Outside consumers can register new document types using [ActsAsHashable::Factory#register.](https://github.com/bluemarblepayroll/acts_as_hashable/blob/master/lib/acts_as_hashable/factory.rb#L31)

**Comment by mattruggio** [comment]
> Most of my comments are fairly minor and several are regarding documentation. One overall thought is the creation of zip files without an enclosing directory. This is a pet peeve of mine. When I extract a zip file which puts its contents in the current directory instead of extracting everything in an enclosing directory, I am irritated about the clutter that was just created. It would be nice if one could specify an enclosing directory for zip files via the blueprint. A similar concern is that even though there is file uniqueness checking at the top level, it would be possible to create a zip file that when extracted, conflicts with other files. For example, modifying your hello_worlds_zip.yaml example:
> 
> ```
> documents:
>   - type: text
>     filename: hello_world_1.txt
>     data: FIRST HELLO WORLD!
>   - type: text
>     filename: hello_world_2.txt
>     data: SECOND HELLO WORLD!
>   - type: zip
>     filename: nested.zip
>     blueprint:
>       documents:
>         - type: text
>           filename: hello_world_1.txt
>           data: THIRD HELLO WORLD!
> ```
> 
> Note that I do not get an error about hello_world_1.txt being a duplicate file name. Typically unzip programs warn a user before overwriting a file when extracting. However, it would probably be handy to have some guardrails for Blueprint authors to prevent them from creating scenarios where the end user of the created files has to make an unexpected decision about duplication.

Let's defer this until we find it becoming an issue.  Technically it does not prevent the library from properly generating the desired files, it only *potentially* affects the user's experience when extracting/downloading files.

File: lib/file_composer/blueprint.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  # The main object model.
+  class Blueprint
+    acts_as_hashable
+
+    attr_reader :documents
+
+    def initialize(documents: [])
+      @documents = Documents.array(documents)
+      filenames  = @documents.map { |a| a.filename.downcase }
+      not_unique = filenames.uniq.length != @documents.length
```
Comment (gerryster) [inline lib/file_composer/blueprint.rb:23]: Nice, this is a lot more clear.

**Review by gerryster — CHANGES_REQUESTED** [review]
Ok, @mattruggio : one very minor copy change idea for you. Also, it looks like the build is failing due to Rubocop. Looks good otherwise!

File: file_composer.gemspec [hunk 6]
```diff
@@ -5,10 +5,10 @@ require './lib/file_composer/version'
 Gem::Specification.new do |s|
   s.name        = 'file_composer'
   s.version     = FileComposer::VERSION
-  s.summary     = 'TBD'
+  s.summary     = 'High-level, pluggable, and declarative API for creating files'
 
   s.description = <<-DESCRIPTION
-    TBD
+    This library provides a YAML-based configuration, called a Blueprint, that allows you to specify text files and zip files and it will create them.  It is pluggable so other document types and storage mediums can be added.
```
[the remark this answers is not in the document]
Comment (mattruggio) [reply file_composer.gemspec:11]: Looked like rubocop-ast published a new minor version and broke compatibility, I locked in the known compatible one for now.

[the remark this answers is not in the document]
Comment (mattruggio) [reply file_composer.gemspec]: Updated README as well.

**Review by gerryster — APPROVED** [review]

**Pull request merged**